R/ETagMiddleware.R

```diff
@@ -0,0 +1,222 @@
+#'                   headers = list("If-None-Match" = "WRONG HASH"))
+#' app$process_request(req)
+#'
+#' # alternatively, you can provide a timestamp in the If-Modified-Since header
+#' # => 304 Not Modified (Can be cached)
+#' modified_since = format(last_modified + 1, "%FT%TZ")
+#' req = Request$new(path = "/example.txt",
+#'                   headers = list("If-Modified-Since" = modified_since))
+#' app$process_request(req)
+#'
+#' # provide both headers: If-None-Match takes precedence
+#' # in this case:
+#' #  - if none match => modified (No cache)
+#' #  - if modified since => NOT MODIFIED (cached)
+#' # => Overall: modified = no cache
+#' modified_since = format(last_modified + 1, "%FT%TZ")
+#' req = Request$new(path = "/example.txt",
+#'                   headers = list("If-None-Match" = "CLEARLY WRONG",
+#'                                  "If-Modified-Since" = modified_since))
+#' app$process_request(req)
+#'
+#' #############################################################################
+#'
+#' # use an alternative hash function (use name of the file)
+#' hash_on_filename = function(x) x
+#' # also use an alternate last_modified time function
+#' always_1900 = function(x) as.POSIXlt("1900-01-01 12:34:56", tz = "GMT")
+#'
+#' app = Application$new(middleware = list(
+#'   ETagMiddleware$new(hash_function = hash_on_filename,
+#'                      last_modified_function = always_1900)
+#' ))
+#' app$add_static(path = "/", file_path = static_dir)
+#'
+#' req = Request$new(path = "/example.txt")
+#' (res = app$process_request(req))
+#'
+#' filename = res$body[["file"]]
+#' req = Request$new(path = "/example.txt",
+#'                   headers = list("If-None-Match" = filename))
+#' app$process_request(req)
+ETagMiddleware = R6::R6Class(
+  classname = "EtagMiddleware",
+  inherit = Middleware,
+  public = list(
+    #' @field hash_function Function that takes an object or file and computes
+    #' the hash of it
+    hash_function = NA,
+    #' @field last_modified_function Function that takes an object or file and
+    #' computes the last time it was modified
+    last_modified_function = NA,
+    #' @description
+    #' Creates ETag middleware object
+    #' @param routes Routes paths to protect.
+    #' @param match How routes will be matched: exact or partial (as prefix).
+    #' @param id Middleware id.
+    #' @param hash_function a function that generates the ETag hash.
+    #' The function takes the body of the response and returns a single
+    #' character. Default is crc32 using [digest::digest].
+    #' @param last_modified_function a function that takes the body of the
+    #' response and returns the last time this was changed. The default is to
+    #' take the mtime (last time the file was modified) if its a file,
+    #' if the body does not contain a file, the current time is returned (
+    #' resulting in no caching)
+    initialize = function(routes = "/", match = "partial",
```

Comment:
Just noted that ETag does not utilize the routes and match argument. If desired, this can easily be added

Reply:
This is solved by [71ebd3f](https://github.com/rexyai/RestRserve/pull/182/commits/71ebd3f4716466212eeb09f355e034cea7a885ec)